File: src/main/java/com/matag/game/turn/action/cast/InstantSpeedService.java
package com.matag.game.turn.action.cast;

import org.springframework.stereotype.Component;

import com.matag.cards.ability.type.AbilityType;
import com.matag.game.cardinstance.CardInstance;

import lombok.AllArgsConstructor;

@Component
@AllArgsConstructor
public class InstantSpeedService {
  public boolean isAtInstantSpeed(CardInstance cardToCast, String playedAbility) {
    if (playedAbility != null) {
      var abilities = cardToCast.getAbilitiesByType(AbilityType.valueOf(playedAbility));
      for (var ability : abilities) {
        if (!ability.getAbility().getSorcerySpeed()) {
          return true;
        }
      }
    }

    if (cardToCast.isInstantSpeed()) {
      return true;
    }

    return false;
  }

}

File: src/main/kotlin/com/matag/game/turn/action/cast/InstantSpeedService.kt
package com.matag.game.turn.action.cast

import com.matag.cards.ability.type.AbilityType
import com.matag.game.cardinstance.CardInstance
import org.springframework.stereotype.Component

@Component
class InstantSpeedService {
    fun isAtInstantSpeed(cardToCast: CardInstance, playedAbility: String?): Boolean {
        if (playedAbility != null) {
            val abilities = cardToCast.getAbilitiesByType(AbilityType.valueOf(playedAbility))
            for (ability in abilities) {
                if (!ability!!.sorcerySpeed) {
                    return true
                }
            }
        }

        if (cardToCast.isInstantSpeed) {
            return true
        }

        return false
    }
}
